fix(components): update to include mode classes on missing components

brandyscarney opened the pull request:
## Short description of what this resolves:
Adds the mode class to the components that don't have it & need it, for example:

```html
<ion-card>
``` 
⬇️⬇️⬇️ 

```html
<!-- in iOS -->
<ion-card class="ios">

<!-- in MD -->
<ion-card class="md">
```

--------------------------------------------------------------------------------

Removed mode-less component classes from the internal CSS & component:

```css
.slides {
  // some CSS
}
```

⬇️⬇️⬇️ 

```css
ion-slides {
  // some CSS
}
```

## Changes proposed in this pull request:

- [x] **Removed the internal component (mode-less) classes from the following (for example, `.card-content`, `.footer`) and, if it was being used to target the element in the CSS, replaced it with the element's tag:**
    - [x] `ion-card-content`
    - [x] `ion-footer`
    - [x] `ion-header`
    - [x] `ion-infinite-scroll-content`
    - [x] `ion-item-group`
    - [x] `ion-list`
    - [x] `ion-picker`
    - [x] `ion-refresher`
    - [x] `ion-slides`
    - [x] `ion-split-pane`
- [x] **Added tests to make sure `item` is added as a class to the following:**
    - [x] `ion-item`
    - [x] `ion-item-divider`
    - [x] `ion-item-group`
- [x] **Added tests to make sure `button` is added as a class to the following:**
    - [x] `ion-button`
    - [x] `ion-back-button`
    - [x] `ion-menu-button`
- [x] **Added tests to verify the following have their mode specific classes (e.g. `card-content-md`) still for internal styling:**
    - [x] `ion-card-content`
    - [x] `ion-footer`
    - [x] `ion-header`
    - [x] `ion-infinite-scroll-content`
    - [x] `ion-item-group`
    - [x] `ion-item-options`
    - [x] `ion-list`
    - [x] `ion-picker`
    - [x] `ion-refresher`
    - [x] `ion-slides`
    - [x] `ion-split-pane`

**Fixes**: #17608

### Components that do not need mode classes :white_check_mark:

- `action-sheet-controller`
- `alert-controller`
- `keyboard-controller`
- `loading-controller`
- `menu-controller`
- `modal-controller`
- `nav-pop`
- `nav-push`
- `nav-set-root`
- `nav`
- `picker-controller`
- `platform`
- `popover-controller`
- `route-redirect`
- `route`
- `router-outlet`
- `router`
- `tab`
- `tabs`
- `toast-controller`
- `virtual-scroll`


### Components lacking mode classes :x:

:white_check_mark: - component has classes added for mode (`ios`, `md`)
:memo: - e2e test checking for classes has been added

| Class              | Test   | Component                   | Type        |
| -------------------| -------| ----------------------------| ------------|
| :white_check_mark: | :memo: | `action-sheet`              | `scoped`    |
| :white_check_mark: | :memo: | `alert`                     | `scoped`    |
| :white_check_mark: | :memo: | `anchor`                    | `shadow`    |
| :white_check_mark: | :memo: | `app`                       |             |
| :white_check_mark: | :memo: | `avatar`                    | `shadow`    |
| :white_check_mark: | :memo: | `back-button`               | `scoped`    |
| :white_check_mark: | :memo: | `backdrop`                  | `shadow`    |
| :white_check_mark: | :memo: | `badge`                     | `shadow`    |
| :white_check_mark: | :memo: | `button`                    | `shadow`    |
| :white_check_mark: | :memo: | `buttons`                   | `scoped`    |
| :white_check_mark: | :memo: | `card-content`              |             |
| :white_check_mark: | :memo: | `card-header`               | `shadow`    |
| :white_check_mark: | :memo: | `card-subtitle`             | `shadow`    |
| :white_check_mark: | :memo: | `card-title`                | `shadow`    |
| :white_check_mark: | :memo: | `card`                      | `scoped`    |
| :white_check_mark: | :memo: | `checkbox`                  | `shadow`    |
| :white_check_mark: | :memo: | `chip`                      | `shadow`    |
| :white_check_mark: | :memo: | `col`                       |             |
| :white_check_mark: | :memo: | `content`                   | `shadow`    |
| :white_check_mark: | :memo: | `datetime`                  | `shadow`    |
| :white_check_mark: | :memo: | `fab-button`                | `shadow`    |
| :white_check_mark: | :memo: | `fab-list`                  | `shadow`    |
| :white_check_mark: | :memo: | `fab`                       |             |
| :white_check_mark: | :memo: | `footer`                    |             |
| :white_check_mark: | :memo: | `grid`                      | `shadow`    |
| :white_check_mark: | :memo: | `header`                    |             |
| :white_check_mark: | :memo: | `icon`                      |             |
| :white_check_mark: | :memo: | `img`                       | `shadow`    |
| :white_check_mark: | :memo: | `infinite-scroll`           |             |
| :white_check_mark: | :memo: | `infinite-scroll-content`   |             |
| :white_check_mark: | :memo: | `input`                     | `scoped`    |
| :white_check_mark: | :memo: | `item-divider`              | `shadow`    |
| :white_check_mark: | :memo: | `item-option`               | `shadow`    |
| :white_check_mark: | :memo: | `item-options`              |             |
| :white_check_mark: | :memo: | `item-sliding`              |             |
| :white_check_mark: | :memo: | `item`                      | `shadow`    |
| :white_check_mark: | :memo: | `item-group`                |             |
| :white_check_mark: | :memo: | `label`                     | `scoped`    |
| :white_check_mark: | :memo: | `list`                      |             |
| :white_check_mark: | :memo: | `list-header`               | `shadow`    |
| :white_check_mark: | :memo: | `loading`                   | `scoped`    |
| :white_check_mark: | :memo: | `menu-button`               | `shadow`    |
| :white_check_mark: | :memo: | `menu-toggle`               | `shadow`    |
| :white_check_mark: | :memo: | `menu`                      | `shadow`    |
| :white_check_mark: | :memo: | `modal`                     | `scoped`    |
| :white_check_mark: | :memo: | `note`                      | `shadow`    |
| :white_check_mark: | :memo: | `picker`                    | `scoped`    |
| :white_check_mark: | :memo: | `picker-column`             |             |
| :white_check_mark: | :memo: | `popover`                   | `scoped`    |
| :white_check_mark: | :memo: | `progress-bar`              | `shadow`    |
| :white_check_mark: | :memo: | `radio-group`               |             |
| :white_check_mark: | :memo: | `radio`                     | `shadow`    |
| :white_check_mark: | :memo: | `range`                     | `shadow`    |
| :white_check_mark: | :memo: | `refresher`                 |             |
| :white_check_mark: | :memo: | `refresher-content`         |             |
| :white_check_mark: | :memo: | `reorder-group`             |             |
| :white_check_mark: | :memo: | `reorder`                   | `shadow`    |
| :white_check_mark: | :memo: | `ripple-effect`             | `shadow`    |
| :white_check_mark: | :memo: | `row`                       | `shadow`    |
| :white_check_mark: | :memo: | `searchbar`                 | `scoped`    |
| :white_check_mark: | :memo: | `segment-button`            | `shadow`    |
| :white_check_mark: | :memo: | `segment`                   | `scoped`    |
| :white_check_mark: | :memo: | `select-option`             | `shadow`    |
| :white_check_mark: | :memo: | `select-popover`            | `scoped`    |
| :white_check_mark: | :memo: | `select`                    | `shadow`    |
| :white_check_mark: | :memo: | `skeleton-text`             | `shadow`    |
| :white_check_mark: | :memo: | `slide`                     |             |
| :white_check_mark: | :memo: | `slides`                    | `shadow`    |
| :white_check_mark: | :memo: | `spinner`                   | `shadow`    |
| :white_check_mark: | :memo: | `split-pane`                |             |
| :white_check_mark: | :memo: | `tab-bar`                   | `shadow`    |
| :white_check_mark: | :memo: | `tab-button`                | `shadow`    |
| :white_check_mark: | :memo: | `text`                      | `shadow`    |
| :white_check_mark: | :memo: | `textarea`                  | `scoped`    |
| :white_check_mark: | :memo: | `thumbnail`                 | `shadow`    |
| :white_check_mark: | :memo: | `title`                     | `shadow`    |
| :white_check_mark: | :memo: | `toast`                     | `shadow`    |
| :white_check_mark: | :memo: | `toggle`                    | `shadow`    |
| :white_check_mark: | :memo: | `toolbar`                   | `shadow`    |

liamdebeasi commented on core/src/components/tabs/test/standalone/e2e.ts:
```diff
@@ -0,0 +1,26 @@
+  compares.push(await page.compareScreenshot(`tab one`));
+
+  tabButton = await page.find('#tab-button-tab-two');
+  await tabButton.click();
```
Do we need to wait for the tab content to become visible? just worried about a slow CI running these

brandyscarney replied:
I basically just moved the vanilla tabs test to be standalone. This is how it works in the basic test so it should work the same here.

liamdebeasi commented on core/src/components/card/test/basic/e2e.ts:
```diff
@@ -1,10 +1,33 @@
+// ios => expect(el).toHaveClass(`card-content-ios`);
+// mode => expect(el).toHaveClass(`card-content-md`);
+// -----------------------------------------------------------------
+async function checkMode(el: E2EElement, name?: string) {
```
can we move this to the generic test utils file? also do you think we need to do this check anywhere else? it might be good to have tests that test the utility functions we write.

liamdebeasi commented on core/src/utils/test/utils.ts:
```diff
@@ -87,3 +89,45 @@ export const waitForFunctionTestContext = async (fn: any, params: any, interval
+  const mode = await el.getProperty('mode');
+
+  // TODO remove
+  console.log(`component: ${el.nodeName} classes: ${el.className}`);
```
Don't forget to remove this

brandyscarney replied:
I must've pushed the commit to remove this while you were reviewing somehow, it's gone now! 🙂

liamdebeasi commented on core/src/utils/test/utils.ts:
```diff
@@ -87,3 +89,45 @@ export const waitForFunctionTestContext = async (fn: any, params: any, interval
+  const mode = await el.getProperty('mode');
+
+  // TODO remove
+  console.log(`component: ${component} classes: ${el.className}`);
```
Don't forget to remove this

brandyscarney commented:
<img width="1245" alt="Screen Shot 2019-03-26 at 2 01 57 PM" src="https://user-images.githubusercontent.com/6577830/55021803-c6952b00-4fcf-11e9-9c01-4ebac1740742.png">

this broke the toolbar/scenario test

brandyscarney commented:
The bug in my last comment is directly related to https://github.com/ionic-team/ionic/issues/17876

brandyscarney commented:
@liamdebeasi Could you take another look at this? I think the classes themselves are all good, I added it to ionicons and updated the dependency. I reverted the change with title because I think trying to remove the mode classes may be a breaking change and I want to investigate that more.
